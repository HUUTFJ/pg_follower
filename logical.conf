wal_level=logical
log_min_messages = debug1