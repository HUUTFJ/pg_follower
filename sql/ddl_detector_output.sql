
SELECT * FROM pg_create_logical_replication_slot('test', 'ddl_detector');

CREATE TABLE foo (id int);

BEGIN;
INSERT INTO foo VALUES (1);
UPDATE foo SET id = 2 WHERE id = 1;
DELETE FROM foo;
COMMIT;

SELECT * FROM pg_logical_slot_get_changes('test', NULL, NULL, 'include-xids', '0');

DROP TABLE foo;
SELECT * FROM pg_drop_replication_slot('test');
