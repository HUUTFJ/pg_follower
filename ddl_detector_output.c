/*-------------------------------------------------------------------------
 *
 * ddl_detector_output.c
 *
 * IDENTIFICATION
 *		ddl_detector/ddl_detector_output.c
 *
 *-------------------------------------------------------------------------
 */

#include "postgres.h"

#include "replication/logical.h"

extern void _PG_init(void);

/* Support routines */
static void output_insert(StringInfo out,
						  Relation relation,
						  ReorderBufferChange *change);
static void output_update(StringInfo out,
						  Relation relation,
						  ReorderBufferChange *change);
static void output_delete(StringInfo out,
						  Relation relation,
						  ReorderBufferChange *change);

/* Callback routines */
static void detector_startup(LogicalDecodingContext *ctx,
							 OutputPluginOptions *options,
							 bool is_init);
static void detector_begin(LogicalDecodingContext *ctx, ReorderBufferTXN *txn);
static void detector_change(LogicalDecodingContext *ctx,
							ReorderBufferTXN *txn,
							Relation relation,
							ReorderBufferChange *change);
static void detector_commit(LogicalDecodingContext *ctx,
							ReorderBufferTXN *txn,
							XLogRecPtr commit_lsn);
static void detector_message(LogicalDecodingContext *ctx,
							 ReorderBufferTXN *txn,
							 XLogRecPtr message_lsn,
							 bool transactional,
							 const char *prefix,
							 Size message_size,
							 const char *message);

static void
detector_startup(LogicalDecodingContext *ctx, OutputPluginOptions *options,
				 bool is_init)
{
	options->output_type = OUTPUT_PLUGIN_TEXTUAL_OUTPUT;
}

static void
detector_begin(LogicalDecodingContext *ctx, ReorderBufferTXN *txn)
{
	OutputPluginPrepareWrite(ctx, true);

	appendStringInfo(ctx->out, "BEGIN;");

	OutputPluginWrite(ctx, true);
}

static void
output_insert(StringInfo out, Relation relation, ReorderBufferChange *change)
{
	appendStringInfoString(out, "INSERTED");
}

static void
output_update(StringInfo out, Relation relation, ReorderBufferChange *change)
{
	appendStringInfoString(out, "UPDATED");
}

static void
output_delete(StringInfo out, Relation relation, ReorderBufferChange *change)
{
	appendStringInfoString(out, "DELETED");
}

static void
detector_change(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
				Relation relation, ReorderBufferChange *change)
{
	switch (change->action)
	{
		case REORDER_BUFFER_CHANGE_INSERT:
			OutputPluginPrepareWrite(ctx, true);
			output_insert(ctx->out, relation, change);
			OutputPluginWrite(ctx, true);
			break;
		case REORDER_BUFFER_CHANGE_UPDATE:
			OutputPluginPrepareWrite(ctx, true);
			output_update(ctx->out, relation, change);
			OutputPluginWrite(ctx, true);
			break;
		case REORDER_BUFFER_CHANGE_DELETE:
			OutputPluginPrepareWrite(ctx, true);
			output_delete(ctx->out, relation, change);
			OutputPluginWrite(ctx, true);
			break;
		default:
			elog(ERROR, "unknown change");
			break;
	}
}

static void
detector_commit(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
				XLogRecPtr commit_lsn)
{
	OutputPluginPrepareWrite(ctx, true);

	appendStringInfo(ctx->out, "COMMIT;");

	OutputPluginWrite(ctx, true);
}

static void
detector_message(LogicalDecodingContext *ctx, ReorderBufferTXN *txn,
				 XLogRecPtr message_lsn, bool transactional,
				 const char *prefix, Size message_size, const char *message)
{

}

/* Public routines */

void
_PG_init(void)
{
    /* NO-OP */
}

/* Specify output plugin callbacks */
void
_PG_output_plugin_init(OutputPluginCallbacks *cb)
{
	AssertVariableIsOfType(&_PG_output_plugin_init, LogicalOutputPluginInit);

	cb->startup_cb = detector_startup;
	cb->begin_cb = detector_begin;
	cb->change_cb = detector_change;
	cb->commit_cb = detector_commit;
	cb->message_cb = detector_message;
}
