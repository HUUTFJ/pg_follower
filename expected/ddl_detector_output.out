SELECT * FROM pg_create_logical_replication_slot('test', 'ddl_detector');
 slot_name |    lsn    
-----------+-----------
 test      | 0/191E408
(1 row)

CREATE TABLE foo (id int);
INFO:  deparse result: CREATE TABLE  public.foo ( id pg_catalog.int4 );
BEGIN;
INSERT INTO foo VALUES (1);
UPDATE foo SET id = 2 WHERE id = 1;
DELETE FROM foo;
COMMIT;
SELECT * FROM pg_logical_slot_get_changes('test', NULL, NULL, 'include-xids', '0');
    lsn    | xid |   data   
-----------+-----+----------
 0/191E478 | 743 | BEGIN;
 0/191F9E8 | 743 | COMMIT;
 0/191F9E8 | 744 | BEGIN;
 0/191F9E8 | 744 | INSERTED
 0/191FA28 | 744 | UPDATED
 0/191FA70 | 744 | DELETED
 0/191FAD8 | 744 | COMMIT;
(7 rows)

DROP TABLE foo;
SELECT * FROM pg_drop_replication_slot('test');
 pg_drop_replication_slot 
--------------------------
 
(1 row)

